sdks/java/core/src/main/java/org/apache/beam/sdk/io/TextIO.java

```diff
@@ -653,31 +684,51 @@ public Bound withFooter(@Nullable String footer) {
       public Bound withWritableByteChannelFactory(
           WritableByteChannelFactory writableByteChannelFactory) {
         return new Bound(name, filenamePrefix, filenameSuffix, header, footer, numShards,
-            shardTemplate, validate, writableByteChannelFactory);
+            shardTemplate, validate, writableByteChannelFactory, filenamePolicy, windowedWrites);
+      }
+
+      public Bound withWindowedWrites() {
+        return new Bound(name, filenamePrefix, filenameSuffix, header, footer, numShards,
+            shardTemplate, validate, writableByteChannelFactory, filenamePolicy, true);
       }
 
       @Override
       public PDone expand(PCollection<String> input) {
-        if (filenamePrefix == null) {
+        if (filenamePolicy == null && filenamePrefix == null) {
+          throw new IllegalStateException(
+              "need to set the filename prefix of an TextIO.Write transform");
+        }
+        if (filenamePolicy != null && filenamePrefix != null) {
           throw new IllegalStateException(
-              "need to set the filename prefix of a TextIO.Write transform");
+              "cannot set both a filename policy and a filename prefix");
+        }
+        org.apache.beam.sdk.io.Write<String> write = null;
+        if (filenamePolicy != null) {
```

Comment:
Somewhere in `expand` (if not compile time) we should check that the filename policy supports windowing OR the input is globally windowed.